### PR TITLE
New generation

### PR DESCRIPTION
**Pull request by jendakol**
* Reworked to new structure
* Finally tagless

There is a lot of changes because it's refactoring of old StorClient. Interesting classes are:
1. StorClient
1. StorBackend
1. storage package object

### CODE REVIEW

File: core/src/main/scala/com/avast/clients/storage/StorageClient.scala [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+
+import scala.language.higherKinds
+
+trait StorageClient[F[_]] extends AutoCloseable {
+  def head(sha256: Sha256): F[Either[StorageException, HeadResult]]
+
+  def get(sha256: Sha256, dest: File = File.newTemporaryFile(prefix = "stor")): F[Either[StorageException, GetResult]]
+}
+
+class DefaultStorageClient[F[_]: Effect](backend: StorageBackend[F]) extends StorageClient[F] {
```
Comment (augi) [inline core/src/main/scala/com/avast/clients/storage/StorageClient.scala]: Maybe `SingleStorageClient`?

File: stor/src/main/scala/com/avast/clients/storage/stor/stor.scala [hunk 2]
```diff
@@ -1,35 +1,5 @@
...
-        client.get(sha256, dest)
-      }
-    }
-
-  }
-
-
-}
 
+package object stor {}
```
Comment (augi) [inline stor/src/main/scala/com/avast/clients/storage/stor/stor.scala]: Is this file required?

File: core/src/main/scala/com/avast/clients/storage/storage.scala [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+
+package object storage {
+
+  implicit class StorageClientOps[F[_]](val client: StorageClient[F]) extends AnyVal {
+    def mapK[G[_]](implicit fToG: F ~> G): StorageClient[G] = {
+      storageClientFunctorK.mapK(client)(fToG)
+    }
+  }
+
+  implicit class StorageBackendOps[F[_]: Monad](val backend: StorageBackend[F]) {
```
Comment (augi) [inline core/src/main/scala/com/avast/clients/storage/storage.scala]: `extends AnyVal`?

File: core/src/main/scala/com/avast/clients/storage/storage.scala [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+
+import better.files.File
+import cats.data.EitherT
+import cats.{~>, Monad}
+import com.avast.scala.hashes.Sha256
+import mainecoon.FunctorK
+
+import scala.language.higherKinds
+
+package object storage {
```
Comment (augi) [inline core/src/main/scala/com/avast/clients/storage/storage.scala]: Why is this class needed? Where is it used pls?

File: stor/src/main/scala/com/avast/clients/storage/stor/StorBackend.scala [hunk 5]
```diff
@@ -0,0 +1,162 @@
...
+            F.pure(Right(HeadResult.NotFound))
+
+          case _ =>
+            resp.bodyAsText.compile.last.map { body =>
+              Left(StorageException.InvalidResponseException(resp.status.code, body.toString, "Unexpected status"))
+            }
+        }
+      }
+    } catch {
+      case NonFatal(e) => F.raiseError(e)
```
Comment (augi) [inline stor/src/main/scala/com/avast/clients/storage/stor/StorBackend.scala]: Why not `F.pure(Left(...))`

Comment (jendakol) [reply stor/src/main/scala/com/avast/clients/storage/stor/StorBackend.scala]: Well, this is kind of a question. Should be throwing an _unknown_ exception considered as fatal failure or as normal state? ;-)

File: core/src/main/scala/com/avast/clients/storage/storage.scala (hunk 4 above)
Comment (jendakol) [reply core/src/main/scala/com/avast/clients/storage/storage.scala]: `import c.a.clients.storage._`

File: core/src/main/scala/com/avast/clients/storage/storage.scala (hunk 3 above)
Comment (jendakol) [reply core/src/main/scala/com/avast/clients/storage/storage.scala]: Nope. The fact that it has `[F[_]: Monad]` causes that there is second param (`implicit val sth: Monad[F]` generated by compiler) thus it could not `extend AnyVal`.

File: stor/src/main/scala/com/avast/clients/storage/stor/stor.scala (hunk 2 above)
Comment (jendakol) [reply stor/src/main/scala/com/avast/clients/storage/stor/stor.scala]: Nope, I will delete it and re-add it if needed.

File: core/src/main/scala/com/avast/clients/storage/StorageClient.scala (hunk 1 above)
Comment (jendakol) [reply core/src/main/scala/com/avast/clients/storage/StorageClient.scala]: That would be misleading name. The client uses single `StorageBackend` instance however it can be in fact _merged_ from two (or more) instances.

File: stor/src/main/scala/com/avast/clients/storage/stor/StorBackend.scala [hunk 6]
```diff
@@ -0,0 +1,162 @@
...
+import org.http4s.headers.`Content-Length`
+import pureconfig.modules.http4s.uriReader
+import pureconfig.{CamelCase, ConfigFieldMapping, ProductHint}
+
+import scala.concurrent.ExecutionContext
+import scala.concurrent.duration.Duration
+import scala.language.higherKinds
+import scala.util.control.NonFatal
+
+trait StorBackend[F[_]] extends StorageBackend[F]
```
[the remark this answers is not in the document]
Comment (jendakol) [reply stor/src/main/scala/com/avast/clients/storage/stor/StorBackend.scala]: There were some things connected to it but now it's not necessary.
